### PR TITLE
Support logits multirank projection & Metal MoE gguf

### PR DESCRIPTION
**Pull request by guoqingbao**

### CODE REVIEW

**Comment by sempervictus** [comment]
If i'm reading this right, we should be able to use this for _grammar constraint masking in GPU broadcast_ instead of logit-by-logit on CPU as we currently do :thinking:

**Comment by guoqingbao** [comment]
> If i'm reading this right, we should be able to use this for _grammar constraint masking in GPU broadcast_ instead of logit-by-logit on CPU as we currently do 🤔

This PR is used to speedup multirank inference and saving vram at the same time. Since we split the lm_head (the largest weight in LLM) into multiple ranks for parallel processing.

**Comment by sempervictus** [comment]
NVFP4 80B coder on 4xSM120 finally breaking the 100T/S decoding speed!
```rust
2026-06-03T17:21:21.196929Z  WARN xinfer::server::server: --- Performance Metrics ---
2026-06-03T17:21:21.196941Z  INFO xinfer::server::server: [1 requests] ⏱️ Prompt tokens: 17 in 0.05s (333.33 t/s)
2026-06-03T17:21:21.196947Z  INFO xinfer::server::server: [1 requests] ⏱️ Decoded tokens: 160 in 1.58s (101.33 t/s)
```
granted its on 17 tokens of input but 101!

The NVFP4 122B q3.5 on 4xV100 is at almost 40:
```rust
2026-06-03T17:21:06.579154Z  WARN xinfer::server::server: --- Performance Metrics ---
2026-06-03T17:21:06.579165Z  INFO xinfer::server::server: [1 requests] ⏱️ Prompt tokens: 17 in 0.14s (122.30 t/s)
2026-06-03T17:21:06.579171Z  INFO xinfer::server::server: [1 requests] ⏱️ Decoded tokens: 554 in 14.19s (39.03 t/s)
```

**Comment by sempervictus** [comment]
Far as VRAM savings - presumably from this PR (i havent looked at this since #373), we now have 5-6% overhead for dynamic buffer allocation at runtime on the V100s with 1M context length in `turbo4` KV:
image

Without this PR it was ~2% and needed `--prefill-chunk-size 4096` to avoid occasional OOM.

**Comment by sempervictus** [comment]
This appears to fix the 397B NVFP4 OOM issue:
```rust
vllm-rs-svc0  | 2026-06-03T18:30:59.826024Z  WARN xinfer::server::server: --- Performance Metrics ---
vllm-rs-svc0  | 2026-06-03T18:30:59.826046Z  INFO xinfer::server::server: [Seq 4] ⏱️ Prompt: 44267 tokens in 10.06s (4398.11 t/s)
vllm-rs-svc0  | 2026-06-03T18:30:59.826057Z  INFO xinfer::server::server: [Seq 4] ⏱️ Decoded: 829 tokens in 13.49s (61.44 t/s)
```

**Comment by guoqingbao** [comment]
> This appears to fix the 397B NVFP4 OOM issue:
> 
> ```rust
> vllm-rs-svc0  | 2026-06-03T18:30:59.826024Z  WARN xinfer::server::server: --- Performance Metrics ---
> vllm-rs-svc0  | 2026-06-03T18:30:59.826046Z  INFO xinfer::server::server: [Seq 4] ⏱️ Prompt: 44267 tokens in 10.06s (4398.11 t/s)
> vllm-rs-svc0  | 2026-06-03T18:30:59.826057Z  INFO xinfer::server::server: [Seq 4] ⏱️ Decoded: 829 tokens in 13.49s (61.44 t/s)
> ```

It's very promising!

**Comment by guoqingbao** [comment]
> Without this PR it was ~2% and needed `--prefill-chunk-size 4096` to avoid occasional OOM.

You mean this PR avoided some OOMs?

**Comment by sempervictus** [comment]
I mean this pr **fixes** the OOM hate i've beeen experiencing with the 397B for weeks! Thank you :grin: 

8K prefill chunk size is back in business on the 397B NVFP4 over 4xSM120:
```rust
vllm-rs-svc0  | 2026-06-03T22:38:52.825609Z  INFO xinfer::core::scheduler: Chunk prefilled 1 seq(s) [31:8192/31061] (8192 total tokens processed)
vllm-rs-svc0  | 2026-06-03T22:38:54.596854Z  INFO xinfer::core::scheduler: Chunk prefilled 1 seq(s) [31:16384/31061] (16384 total tokens processed)
vllm-rs-svc0  | 2026-06-03T22:38:56.423354Z  INFO xinfer::core::scheduler: Chunk prefilled 1 seq(s) [31:24576/31061] (24576 total tokens processed)
vllm-rs-svc0  | 2026-06-03T22:38:57.872954Z  WARN xinfer::core::scheduler: Chunk prefill finished for 1 seq(s) [31] (31061 total tokens)
```

**Comment by guoqingbao** [comment]
> Far as VRAM savings - presumably from this PR (i havent looked at this since #373), we now have 5-6% overhead for dynamic buffer allocation at runtime on the V100s with 1M context length in `turbo4` KV:

I increased default kv-fraction from 0.5 to 0.6, meaning more kvcache and mamba cache allocated.

**Comment by sempervictus** [comment]
> I increased default kv-fraction from 0.5 to 0.6, meaning more kvcache and mamba cache allocated.

I dont think that's the cause - i use `--yarn-scaling-factor` and `--max-model-len` parameters (with `--num-seqs 1` to maximize the context) which (i believe) bypasses the kv faction logic since it knows how much space each token should use and how many tokens worth of space to preallocate. In recent weeks the prefill overhead was OOMing (and in some cases _gdn state capture_) even with a fairly large chunk of overhead left atop KVCache which presumably was that lm head being fully on each GPU and allocating memory for all the vocab each had.

**Comment by guoqingbao** [comment]
> I mean this pr **fixes** the OOM hate i've beeen experiencing with the 397B for weeks! Thank you 😁
> 
> 8K prefill chunk size is back in business on the 397B NVFP4 over 4xSM120:
> 
> ```rust
> vllm-rs-svc0  | 2026-06-03T22:38:52.825609Z  INFO xinfer::core::scheduler: Chunk prefilled 1 seq(s) [31:8192/31061] (8192 total tokens processed)
> vllm-rs-svc0  | 2026-06-03T22:38:54.596854Z  INFO xinfer::core::scheduler: Chunk prefilled 1 seq(s) [31:16384/31061] (16384 total tokens processed)
> vllm-rs-svc0  | 2026-06-03T22:38:56.423354Z  INFO xinfer::core::scheduler: Chunk prefilled 1 seq(s) [31:24576/31061] (24576 total tokens processed)
> vllm-rs-svc0  | 2026-06-03T22:38:57.872954Z  WARN xinfer::core::scheduler: Chunk prefill finished for 1 seq(s) [31] (31061 total tokens)
> ```

Fabulous!
